### ISSUE TITLE
Latest update (v4.6.1) causes BO to not work anymore.

### ISSUE THREAD

**Issue by kmorgen** [PrestaShopCorp/ps_mbo#697]
**Describe the bug and add attachments:**
Just auto updated to latest version 'PrestaShop Marketplace in your Back Office' (v4.6.1), and it caused BO to crash completely,

Managed to get the following error messages:

![PS-error](https://github.com/PrestaShopCorp/ps_mbo/assets/4300688/97ed5fe1-7351-4961-82a7-5cab0a19b372)

Uninstalling the module caused the entire shop to shop working including Front.

![ps_mbo](https://github.com/PrestaShopCorp/ps_mbo/assets/4300688/cd0b9b87-328e-4601-ba19-5f2bac434f61)

Installed the previous version V4.5.3 and the shop could apparently run without Debug mode again. Did another update, and it completely crashed the shop again!

**Steps to reproduce**
Auto Update the Marketplace module from 4.5.3 to 4.61 under Modules gave the error message afterwards. Then Modules page, as well as other pages like Performance/Cache page etc no longer works.

Uninstalling the module caused the shop to crash in FO also.
Installing V4.5.3 again seem to address the error related issue, but not sure if everything works again (did a full restore from backup instead)

**PrestaShop version(s) where the bug happened**
8.1.2

**PHP version(s) where the bug happened**
8.1.26

**If your bug is related to a module, specify its name and its version**
ps_mbo 4.6.1

**Comment by kmorgen** [comment]
Hell @intraordinaire 

Thanks for your feedback!

I just tried with a new testshop.
When upgrading the module I still get an red error message at the end saying something like: Module could not perform upgrade undefined".  Then afterwards the debug error screen appears.

I then tried to delete the mentioned cache folder. I could then refresh the page and get access again! 
However the module was still mentioned as V5.5.3.

I then changed the cache settings to 'Force Compile', and then tried to upgrade module again. The second time it then upgraded to 4.6.1 and no site crash. However still received this message in modules afterwards:

![image](https://github.com/PrestaShopCorp/ps_mbo/assets/4300688/e5ea6429-0064-4019-9933-a06b86b29c50)

So it does seem cache related, but just a bit weird the module will fail the upgrade. It seems it does not finish, and thereby not clearing the cache itself.

Anyway, the testsite seems to run now. So maybe I will dare to try on live shops later...

**Comment by intraordinaire** [comment]
@kmorgen 

We did some investigations yesterday on this.

We manage to reproduce the error with PrestaShop 8.0.0 or 8.0.1.
PrestaShop 8.0.2 came with improvements about module updates, and we did not manage to crash the shops.

We still have some troubles updating the module on newer version >= 8.0.2, but for now, we can not do much more.
I explain myself.

The process to update a module is the following : 
 - PrestaShop get the updated source (through download or by uploading a zip)
 - Replace old files with new ones
 - Run upgrade scripts
 
The trouble we're having with our module, is that we are responsible to manage external modules installation and/or updates.
So, our module is already loaded in PHP process when replacing files.
Then, when trying to run upgrade script, upgrade the version in DB, and so on, it may cause problem because there is different version of files/services between the ones that are already load, and the one that are about to be...
With this problem, we have to update the module twice to get the update working correctly.
First time, it displays an error, but the files are correctly replaced. Second, it will not touch the files, and just run the upgrade scripts.

PrestaShop 8.1.2 with ps_mbo ready to be updated to 4.6.1
![Capture d’écran 2023-11-30 à 10 13 43](https://github.com/PrestaShopCorp/ps_mbo/assets/1721887/8f781f17-702f-4675-bfe6-b943a7f807ef)

First update failed 😢 
![Capture d’écran 2023-11-30 à 10 13 56](https://github.com/PrestaShopCorp/ps_mbo/assets/1721887/35a3fb74-3bf9-4b24-a225-95385e038475)

Second update success 🎉 
![Capture d’écran 2023-11-30 à 10 14 13](https://github.com/PrestaShopCorp/ps_mbo/assets/1721887/8adb5edb-e4d2-4e94-a5ad-1726149f0828)

Shop still up, and module well updated
![Capture d’écran 2023-11-30 à 10 14 30](https://github.com/PrestaShopCorp/ps_mbo/assets/1721887/8a5ec9dd-ae65-43c3-8e9a-d87dca38390f)


And, if I reload the page between the upgrades, the shop is still up and the module is still displayed as 4.5.3, even if the source code is in 4.6.1.
After several tests with versions of PrestaShop >= 8.0.2 it seems to "work" like this.

If I test with a shop in PrestaShop 8.0.0 or 8.0.1, I did have the broken shop every time 😞 

We are still trying to find a way to get it working properly.
We may use some update mechanism we have developed in our module.
Do the update process in two times, first download the source and exit, and then, in another process, run the upgrade, which seems to work like a charm.

In conclusion, for now : 
- We have removed the compatibility of PS < 8.0.2 for this version
- We are still working to fix this
- We will update the readme of the module to add a "Known bugs" section to explain this behavior.

In the meantime (not an ideal solution), to use the last version of the module, you can also uninstall it, and then, install the new one. 
This will not break anything, as it's the update process who is breaking things here. The module itself works well.

Again, thanks for your feedback, really important to us, and don't hesitate if something here is unclear.

**Comment by kmorgen** [comment]
Hi @intraordinaire 

Thank you for your feedback, I will study it.

Just to clarify, when you mention 8.0.2 you actually mean the latest 8.1.2, right? (which I am running too)

So does this also explain, why many modules in general need to be upgraded 2 times?

**Comment by intraordinaire** [comment]
> So does this also explain, why many modules in general need to be upgraded 2 times?

It may, yes ! We experienced that with our module, because he is loaded "high" in the PS process. But we may not be alone to do that.

> Just to clarify, when you mention 8.0.2 you actually mean the latest 8.1.2, right? (which I am running too)

Actually, yesterday morning I was meaning 8.0.2 but after our tests of yesterday, I changed my mind.

So, yesterday, we go through a whole day of testing/debugging updates in different PS versions.

It appears that, indeed, before 8.0.2, it's very complicated to do something clean.

For version after 8.0.2 (so in your case with the 8.1.2), it "works" well in 2 times BUT not all the time... Sometimes, and actually, we don't know why, the CacheClearer of Symfony is not deleting anything.
We found something related to this [here](https://github.com/doctrine/annotations/issues/186).

It seems to be better in Symfony 5, but we can't update the Symfony version like this in PS8 😄
It will be done in PrestaShop 9 🎉 

We may have a solution for this :
- for now, remove our module from the available updates.
- Publish a version 4.5.4 without updating services, but with a cleaner way to delete the cache when we are proceeding module downloads.
- Make this update available through another channel, directly in the marketplace, so we'll make it with our mechanism, in two times.
- Allow user to update to 4.6 and more, only if they have this "cleaner" version.

We'll keep you updated on this.

The bug with the account email will be fixed in the 4.7.0, but a bit delayed due to this problem. (planned to release it this week initially).

Last but not least, on all our tests yesterday we do check the Marketplace page, and it worked well.
Do you have errors in the browser console ?

**Comment by intraordinaire** [comment]
Hi @kmorgen 

I hope you had a great time on this EoY, and I wish you all the best for this new year.

We released updates as explained last month.
Have you tried it ?
You can, normally, see that an update is available directly in the marketplace.

Hope it solves the problem !

**Comment by intraordinaire** [comment]
As we released a new version that should fix this behavior, I close the issue.
Feel free to reopen if needed.

Again, thanks for your feedback !
